### PR TITLE
Allow plugins to have custom entities

### PR DESCRIPTION
**Pull request by BSFishy**
This is a WIP to allow plugins to have custom entities. As of now, it is not quite straightforward and has errors with saving and what not. I will tell you when it works, and I have it working with my plugin.

### CODE REVIEW

**Comment by LousyLynx** [comment]
I forget, where is that used?

**Comment by aramperes** [comment]
https://github.com/GlowstoneMC/Glowstone/pull/404/files#diff-8769715641781a189890669ce58619a0R110

Say you have a custom entity that extends GlowPig, when you register a new store (e.g. CustomPigStore) it will override the default's GlowPig store, which means that all pigs on the server will be loaded with the custom data you are assigning. This might or might not be what is intended when registering new entities.

**Comment by BSFishy** [comment]
Wouldn't we just technically use the idTable then? That would remove the need for the classTable, and allow plugin developers to make custom entities. Or are the other consequences that I don't know about?

**Comment by BSFishy** [comment]
I know in the original NMS it allowed you to use your own enum. I suppose that we could make that happen and have a specific format, with specific methods. If the enum does not have those methods, it throws an error. That way the developer can program it any way they want, and it still works on the server despite the many ways to do it.

**Comment by BSFishy** [comment]
That saves the type as a string rather than an EntityType itself. Also, there is a constructor that takes in the string. It is a string because that is the only thing from EntityType that it is using.

**Comment by BSFishy** [comment]
That should be about it. My entity is registering and saving, and the other entities are working like they were before I started messing with stuff. It compiles and works properly.

**Comment by aramperes** [comment]
I will test this when I have time today

**Comment by aramperes** [comment]
Functionality seems to be alright. This is the code I used to spawn a custom sheep that is always red, and can become dank:

**CustomEntityPlugin**
```java
public class CustomEntityPlugin extends JavaPlugin implements Listener {
    @Override
    public void onLoad() {
        EntityStorage.bind(new CustomSheepStore()); // bind the store before loading the world
    }

    @Override
    public void onEnable() {
        getServer().getPluginManager().registerEvents(this, this);
    }

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        GlowWorld world = (GlowWorld) event.getPlayer().getWorld();
        world.spawn(event.getPlayer().getLocation(), CustomSheep.class, CreatureSpawnEvent.SpawnReason.CUSTOM); // Needs to use method GlowWorld#spawn(location, <class extends GlowEntity>, spawnReason) to force the use of custom entities
    }
}
```

**CustomSheep**
```java
public class CustomSheep extends GlowSheep {
    private boolean dank;

    public CustomSheep(Location location) {
        super(location);
        setColor(DyeColor.RED);
    }

    public boolean isDank() {
        return dank;
    }

    public void setDank(boolean dank) {
        this.dank = dank;
        setColor(dank ? DyeColor.GREEN : DyeColor.RED);
    }

    @Override
    public boolean entityInteract(GlowPlayer player, InteractEntityMessage message) {
        ItemStack holding = InventoryUtil.itemOrEmpty(player.getItemInHand());
        if (holding.getType() == Material.LONG_GRASS && holding.getDurability() == 2 && !isDank()) {
            setDank(true);
            world.spawnParticle(Particle.SMOKE_LARGE, getLocation(), 1);
            // consume
            holding.setAmount(holding.getAmount() - 1);
            if (holding.getAmount() == 0) holding = InventoryUtil.createEmptyStack();
            player.setItemInHand(holding);
            return true;
        }
        return false;
    }
}
```

**CustomSheepStore**

```java
public class CustomSheepStore extends AnimalStore<CustomSheep> {
    public CustomSheepStore() {
        super(CustomSheep.class, "custom_sheep");
    }

    @Override
    public void save(CustomSheep entity, CompoundTag tag) {
        super.save(entity, tag);
        if (entity.isDank()) {
            tag.putBool("Dank", true);
        }
    }

    @Override
    public void load(CustomSheep entity, CompoundTag tag) {
        super.load(entity, tag);
        if (tag.isByte("Dank")) {
            entity.setDank(tag.getBool("Dank"));
        }
    }
}
```

edit: added danks

**Comment by mastercoms** [comment]
I think it would be better to have some sort of wrapper class you can use to extend entities. That way we don't have to use strings instead of the enum.

**Comment by aramperes** [comment]
It should also be possible to override/register new loot data for custom entities. Maybe look into if you can work something out, unless you think this is out of the scope of this PR, then it can be added later on.

**Comment by aramperes** [comment]
Some additions I suggest to your PR: https://github.com/LousyLynx/Glowstone/pull/1

File: src/main/java/net/glowstone/command/SummonCommand.java [hunk 1]
```diff
@@ -88,19 +91,24 @@ public boolean execute(CommandSender sender, String commandLabel, String[] args)
...
-            EntityRegistry.getEntity(type).getConstructor(Location.class);
-        } catch (NoSuchMethodException e) {
-            if (sender != null) sender.sendMessage(ChatColor.RED + "The entity '" + type.getName() + "' cannot be summoned.");
+        if (entityType != null && EntityRegistry.getEntity(entityType) == null) {
+            if (sender != null)
+                sender.sendMessage(ChatColor.RED + "The entity '" + entityType.getName() + "' is not implemented yet.");
+            return false;
+        } else if (entityType == null && !EntityRegistry.isCustomEntityRegistered(type)) {
+            if (sender != null)
+                sender.sendMessage(ChatColor.RED + "The entity '" + type + "' is does not exist.");
```
Comment (aramperes) [inline src/main/java/net/glowstone/command/SummonCommand.java]: "is does not exist" - typo, my bad.

**Review by aramperes — APPROVED** [review]
I'm ok with this, we can always adjust the API if we need to in the future.

**Review by mastercoms — APPROVED** [review]
I would prefer using the EntityType enum, but I think it would be convoluted to have it all work, so I'm fine with this.

**Comment by aramperes** [comment]
Documentation relative to these additions: https://github.com/GlowstoneMC/Glowstone/wiki/Creating-custom-entities

**Pull request merged**